### ISSUE TITLE
ResourceMonitor default waiting values

### ISSUE THREAD

**Issue by EvGe22** [allegroai/clearml#208]
Hi.

Trains version 0.16.1
Python version 3.6.10

My data preprocessing happens after I have initialized the Trains Task object. In the experiment I am currently running, It takes it approx 2.5 minutes to complete it (other experiments might take up to tens of minutes to do this). 
While it's happening, I get a message that Trains Could not detect iteration reporting and it's gonna revert to seconds from the start. But when the first actual batch of the training is completed(right after the processing was done), I do a `logger.report_scalar` with an iteration, resulting in Trains reverting back to iteration based reporting. 

The loss and my custom metrics are being reported as expected - starting from iteration 0, everything is good. 
BUT, the automatic system metrics (CPU, GPU, Memory) are reported incorrectly. My iteration reporting sends an update every % of the training epoch done. Thus, 1 epoch - 100 iterations. The first automatic report is sent approximately 3 min - 3 min 10 sec after the Trains object init and it's the 182nd or 183rd iteration from the beginning.

If, however, I remove most of my training data, so that preprocessing takes < 5-10 secs, the automatic reporting works as intended and I do not even get any warning about not detecting the iteration reporting.

Thanks.

UPD. I have removed the workaround because it does not seem to really work. Posted too soon. Will update it upon further experiments.

UPD 2. Okay, I have found out that the is a ResourceMonitor that has a default waiting parameters [here](https://github.com/allegroai/trains/blob/master/trains/utilities/resource_monitor.py#L23-L24). I have set the automatic resource reporting of the Task.init to False and have created my own new ResourceMonitor with the desired parameters. This required some source-code digging and wasn't completely obvious.
My suggestion is to have the ability to have values be taken from the config file and set to default only if they are empty. Thus, providing a user with a quick and easy change for the waiting period based on his task.

**Comment by bmartinn** [comment]
Hi @EvGe22 
Kudos on the workaround !
I think you are absolutely correct, this default value should be configurable.
Actually I think it makes sense to have the ability to configure it from code.
After all, one experiment could have one value and another a different value, both running on the same machine...
I'm thinking something like:
```python
Task.init('examples', 'delay resource monitor',  auto_resource_monitoring={'wait_for_first_iteration_to_start_sec': 1800})
```

WDYT?

**Comment by EvGe22** [comment]
> After all, one experiment could have one value and another a different value, both running on the same machine...

Completely agree on this. However, there are few things that I do not really like about passing arguments like a dict. 
1. Error-prone - user might mistype the argument name and will have to run the script to see the error
2. Inability to see all of the available parameters using IDE hints, autocomplete.
3. No type checking/code linting.
4. We are limited to only one exact ResourceMonitor implementation.

My suggestion would be to make something along the lines of:
```python
class ResourceMonitor(object):
    _title_machine = ':monitor:machine'
    _title_gpu = ':monitor:gpu'

    def __init__(self, sample_frequency_per_sec=None, report_frequency_sec=None,
                 first_report_sec=None, wait_for_first_iteration_to_start_sec=None,
                 max_wait_for_first_iteration_to_start_sec=None, report_mem_used_per_process=None):
        self._task = None
        self._sample_frequency = sample_frequency_per_sec or config.get("sample_frequency_per_sec", 2.0)
        self._report_frequency = report_frequency_sec or config.get("report_frequency_sec", 30.0)
        self._first_report_sec = first_report_sec or config.get("first_report_sec", report_frequency_sec)
        self._wait_for_first_iteration = wait_for_first_iteration_to_start_sec or config.get(
            "wait_for_first_iteration_to_start_sec", 180.0)
        self._max_check_first_iteration = max_wait_for_first_iteration_to_start_sec or config.get(
            "max_wait_for_first_iteration_to_start_sec", 1800.0)
        ...

    def observe_task(self, task):
        self._task = task
        # optioanlly, also have this here
        if not self._gpustat:
            self._task.get_logger().report_text('TRAINS Monitor: GPU monitoring is not available')

...

class Task(_Task):
...

    @classmethod
    def init(
        cls,
        project_name=None,  # type: Optional[str]
        task_name=None,  # type: Optional[str]
        task_type=TaskTypes.training,  # type: Task.TaskTypes
        reuse_last_task_id=True,  # type: Union[bool, str]
        continue_last_task=False,  # type: Union[bool, str]
        output_uri=None,  # type: Optional[str]
        auto_connect_arg_parser=True,  # type: Union[bool, Mapping[str, bool]]
        auto_connect_frameworks=True,  # type: Union[bool, Mapping[str, bool]]
        resource_monitor=,  # type: ResourceMonitoring
    ):
        ...
        if config.get('auto_resource_monitoring', True) and resource_monitor is None:
            resource_monitor = ResourceMonitor(). # this will get the default arguments from the config file
        if resource_monitor is not None:
            resource_monitor.observe_task(task)
            ...
```

This will create a few more ifs in the code, but the user will be able to explicitly pass a configured or even custom ResourceMonitor to the task (and vice versa). This creates more possibilities for customization is more explicit.

I didn't dig much deeper into the code, so I might not be aware of some additional problems this might create. Would be glad to hear your feedback!

**Comment by bmartinn** [comment]
Hi @EvGe22 
Yes I agree with you on argument visibility (also on type checking as a derivative)
That said, we try not to break the interface (or at least as little as possible), and I suspect not a lot of users would actually inherit and change internal classes like the ResourceMonitor...

How about we just add a function to the Task object, something like:
```python
task.set_resource_monitor_iteration_timeout(seconds_from_start=1800)
```
This way we get get visibility and proper type checking (also better `trains-agent` support, as this will be set in remote execution as well as on local execution)

BTW:
From the ResourceMonitor perspective, you can change the timeout setting even after starting the monitoring services, see [here](https://github.com/allegroai/trains/blob/519677d987a2efeeef5b9f4bc256a757d1823c96/trains/utilities/resource_monitor.py#L107)

WDYT?

**Comment by EvGe22** [comment]
Yeah. Breaking the current interface is not a very good idea, agreed. But, the function seems a bit too specific. What if the user might want to change something other than the iteration timeout? Maybe the names of the graphs or the frequency of the readings. 
After we have exchanged a few ideas, I'm now beginning to think the easiest solution would be to just highlight the possibility to modify the ResourceMonitor in a way that I did and not change the interface whatsoever. Also, since I am the first to create an issue about it, this is probably not a very common problem in general. 

What if we maybe add a small note or a page in the docs? Not sure what exactly it has to say, however.

**Comment by bmartinn** [comment]
> What if the user might want to change something other than the iteration timeout? Maybe the names of the graphs or the frequency of the readings.

Interesting point... I'm thinking both?!
1. Introduce a function to set the timeout, as this is probably the first thing a user might want to do, and this is the simplest interface I can think of.
2. Allow a user to pass an instance or class, of a ResourceMonitor to the `auto_resource_monitoring` argument. This way we do not break the interface, but also allow a user to extend it.

WDYT?

**Comment by EvGe22** [comment]
1. There are no other objective downsides to this that I can think of, so why not.
2. The ResourceMonitor has the task as a required positional argument. So, we cannot pass an instance there. A class, however, sounds good. I personally wouldn't change the type of the `auto_resource_monitoring`. If we want such functionality, IMO it's better to add a separate argument like `resource_monitoring_type` or `resource_monitoring_class`. It's just having a type of `Union[Type[ResourceMonitor], bool]` seems kinda strange.

**Comment by bmartinn** [comment]
@EvGe22 
2. Yes you are correct, class type it is  :) 

I'll update here once an RC is out

**Comment by bmartinn** [comment]
Hi @EvGe22 
Good news 0.16.2 is out with `task.set_resource_monitor_iteration_timeout`, feel free to check it out :)
```
pip install trains==0.16.2
```

**Comment by EvGe22** [comment]
Cool! Thanks, will try it today

**Comment by bmartinn** [comment]
@EvGe22 can I close this issue ? (assuming everything works as expected)

**Comment by EvGe22** [comment]
Sorry. Was busy with other tasks and did not have time to test this out. I will get back to you as soon as I get the results

**Comment by allegroai-git** [comment]
@EvGe22 please reopen if this is still an issue
